psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -0,0 +1,211 @@
+            argsAtPositions.add(atomArgs[index]);
+        }
+        return argsAtPositions;
+    }
+
+    @Override
+    public void compute(TrainingMap trainingMap) {
+        if (defaultPredicate == null) {
+            throw new UnsupportedOperationException("RankingEvaluators must have a default predicate set (through config).");
+        }
+
+        compute(trainingMap, StandardPredicate.get(defaultPredicate));
+    }
+
+    @Override
+    public void compute(TrainingMap trainingMap, StandardPredicate predicate) {
+        labelMap = trainingMap.getLabelMap();
+        rankMap = new HashMap<GroundAtom, Integer>();
+
+        int predicateArity = predicate.getArity();
+        List<Integer> entityIndexes = new ArrayList<Integer>(predicateArity - categoryIndexes.size());
+
+        for (int index = 0; index < predicateArity; index++) {
+            if (categoryIndexes.contains(index)) {
+                continue;
+            } else {
+                entityIndexes.add(index);
+            }
+        }
+
+        // We keep a map from entities to Lists of all GroundAtoms which contain that entity.
+        Map<List<Constant>, List<GroundAtom>> sortedAtoms = new HashMap<List<Constant>, List<GroundAtom>>();
+
+        // Partition the predicted atoms into Lists
+        // according to what entity they contain.
+        for (GroundAtom atom : trainingMap.getAllPredictions()) {
+            if (atom.getPredicate() != predicate) {
+                continue;
+            }
+
+            List<Constant> entity = getArgsAtPositions(atom, entityIndexes);
+
+            if (!sortedAtoms.containsKey(entity)) {
+                sortedAtoms.put(entity, new ArrayList<GroundAtom>());
+            }
+
+            sortedAtoms.get(entity).add(atom);
+        }
+
+        // Sort these Lists and populate the rankMap.
+        for (List<Constant> entity : sortedAtoms.keySet()) {
+            Collections.sort(sortedAtoms.get(entity));
+            for (GroundAtom atom : sortedAtoms.get(entity)) {
+                rankMap.put(atom, sortedAtoms.get(entity).indexOf(atom) + 1);
+            }
+        }
+    }
+
+    /**
+     * Returns the mean reciprocal rank of target atoms
+     * that have a corresponding truth atom with a value above
+     * the specified threshold.
+     */
+    public double mrr() {
+        // These are the numerator and denominator of the MRR.
+        // recRankSum keeps a running sum of reciprocal ranks,
+        // and rankedAtomCount keeps track of how many
+        // atoms are being ranked in this evaluation.
+        double recRankSum = 0.0;
+        int rankedAtomCount = 0;
+
+        for (Map.Entry<GroundAtom, Integer> entry : rankMap.entrySet()) {
+            // If the target atom's corresponding truth value is below
+            // the threshold or it doesn't have a match truth atom, we skip it.
+            if (!labelMap.containsKey(entry.getKey()) || labelMap.get(entry.getKey()).getValue() < threshold) {
```

Comment:
I don't think the threshold makes sense in this context.
What if all the values are under 0.5 (the default value)?
With several items and a functional constraint, this is a very strong possibility.

I don't think it even make senses for this evalautor at all.